shap/benchmark/framework.py

```diff
@@ -0,0 +1,110 @@
+    scores['metrics'].append(metric)
+    scores['values'][metric] = [x, y, auc] 
+
+def get_benchmark(model, X, y, explainer, masker, metrics, exp_num=1, *args):
+    # convert dataframes
+    if safe_isinstance(X, "pandas.core.series.Series") or safe_isinstance(X, "pandas.core.frame.DataFrame"):
+        X = X.values
+    if safe_isinstance(masker, "pandas.core.series.Series") or safe_isinstance(masker, "pandas.core.frame.DataFrame"):
+        masker = masker.values
+    
+    # in case the explainer doesn't have a name
+    try: 
+        name = explainer.name 
+    except: 
+        name = 'explainer' + str(exp_num) 
+        exp_num += 1 
+
+    # record scores per metric 
+    scores = {'name': name, 'metrics': list(), 'values': dict()}
+    for sort_order in metrics['sort_order']:
+        for perturbation in metrics['perturbation']:
+            score_function = lambda true, pred: np.mean(pred)
+            update(model, X, y, explainer, masker, sort_order, score_function, perturbation, scores)
+
+    return scores 
+
+def get_metrics(benchmarks, selection):
+    # select metrics to plot using selection function
+    explainer_metrics = set()
+    for explainer in benchmarks: 
+        scores = benchmarks[explainer]
+        if len(explainer_metrics) == 0: explainer_metrics = set(scores['metrics'])
+        else: explainer_metrics = selection(explainer_metrics, set(scores['metrics']))
+    
+    return list(explainer_metrics)
+
+def trend_plot(benchmarks):
+    explainer_metrics = get_metrics(benchmarks, lambda x, y: x.union(y))
+
+    # plot all curves if metric exists 
+    for metric in explainer_metrics:
+        plt.clf()
+
+        for explainer in benchmarks: 
+            scores = benchmarks[explainer]
+            try: 
+                x, y, auc = scores['values'][metric]
+                plt.plot(x, y, label='{} - {}'.format(round(auc, 3), explainer))
+            except: 
+                pass 
+
+        plt.title(metric)
+        plt.legend()
+        plt.show()
+    
+def compare_plot(benchmarks):
+    explainer_metrics = get_metrics(benchmarks, lambda x, y: x.intersection(y))
+    explainers = list(benchmarks.keys())
+
+    # dummy start to evenly distribute explainers on the left 
+    # can later be replaced by boolean metrics 
+    aucs = dict()
+    for i in range(len(explainers)): 
+        explainer = explainers[i]
+        aucs[explainer] = [i/(len(explainers)-1)] 
+
+    # normalize per metric
+    for metric in explainer_metrics: 
+        max_auc, min_auc = -float('inf'), float('inf')
+
+        for explainer in explainers: 
+            scores = benchmarks[explainer] 
+            _, _, auc = scores['values'][metric]
+            min_auc = min(auc, min_auc)
+            max_auc = max(auc, max_auc)
+        
+        for explainer in explainers: 
+            scores = benchmarks[explainer] 
+            _, _, auc = scores['values'][metric]
+            aucs[explainer].append((auc-min_auc)/(max_auc-min_auc))
```

Comment:
suggestion:
maybe just store
```
             aucs[explainer].append(auc) 
             min_auc = min(auc, min_auc)
             max_auc = max(auc, max_auc)
```
then in the loop below:
```
    auc_diff = max_auc-min_auc
    for explainer in explainers: 
        plt.plot(np.linspace(0, 1, len(explainer_metrics)+1), (aucs[explainer]-min_auc)/auc_diff, '--o')
```
and remove one of the loops above:
```
       for explainer in explainers: 
            scores = benchmarks[explainer] 
            _, _, auc = scores['values'][metric]
            min_auc = min(auc, min_auc)
            max_auc = max(auc, max_auc)
```

Reply:
Agreed. Will change to this

Reply:
Still seems like this can be made simpler, as the lines 
```python
        for explainer in explainers: 
            scores = benchmarks[explainer] 
            _, _, auc = scores['values'][metric]
```
are exactly repeated. Something like:
```python
    for metric in explainer_metrics:
        auc_list = [benchmarks[explainer]['values'][metric][2] for explainer in explainers]
        min_auc = min(auc_list)
        max_auc = max(auc_list)
        auc_diff = max_auc - min_auc
        for auc, explainer in zip(auc_list, explainers): 
            aucs[explainer].append((auc-min_auc)/auc_diff)
```
should be faster and reduces the duplication